src/stochastic_swap.rs

```diff
@@ -0,0 +1,257 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+// Needed to pass shared state between functions
+// closures don't work because of recurssion
+#![allow(clippy::too_many_arguments)]
+
+use std::sync::RwLock;
+
+use hashbrown::HashSet;
+
+use ndarray::prelude::*;
+use numpy::{PyReadonlyArray1, PyReadonlyArray2};
+use rayon::prelude::*;
+
+use pyo3::prelude::*;
+use rand::prelude::*;
+use rand_distr::{Distribution, Normal};
+use rand_pcg::Pcg64Mcg;
+
+use crate::edge_collections::EdgeCollection;
+use crate::nlayout::NLayout;
+
+#[inline]
+fn compute_cost(
+    dist: &ArrayView2<f64>,
+    layout: &NLayout,
+    gates: &[usize],
+    num_gates: usize,
+) -> f64 {
+    (0..num_gates)
+        .map(|kk| {
+            let ii = layout.logic_to_phys[gates[2 * kk]];
+            let jj = layout.logic_to_phys[gates[2 * kk + 1]];
+            dist[[ii, jj]]
+        })
+        .fold(0.0, |a, b| a + b)
+}
```

Comment:
This is minor, but something I was thinking about.  Given floating-point maths, `fold` can't assume that the addition is associative, so it can't do fast-math-like optimisations like turning the fold into a threaded reduction.  Given that this seems to be the worker routine of the loop (by eye - I did exactly zero profiling), is it worth us looking at that for large sizes?

The addition technically still isn't associative, but this shouldn't be a situation where we care about small differences - anything with a coupled link large enough to really screw with the calculation results will never be the best solution anyway, and there can't be catastrophic cancellation because the values should all be positive.

Reply:
It is something I've thought about and why I switched from the for loop in the original cython to the iterator based approach here. My concern with doing a multithreaded reduce here was we're already in a parallel context when this is run so I was worried we just be introducing more overhead.

Basically if you do:

```suggestion
#[inline]
fn compute_cost(
    dist: &ArrayView2<f64>,
    layout: &NLayout,
    gates: &[usize],
    num_gates: usize,
) -> f64 {
    (0..num_gates)
        .into_par_iter()
        .map(|kk| {
            let ii = layout.logic_to_phys[gates[2 * kk]];
            let jj = layout.logic_to_phys[gates[2 * kk + 1]];
            dist[[ii, jj]]
        })
        .reduce(|| 0.0, |a, b| a + b)
}
```

it becomes a parallel reduction. My other thought here was if there was maybe some simd operation we could use here too

As for the fp error here, yeah I agree. Honestly the typing was a bit fast and loose in the cython around this call. The distance matrix here is always an integer (although the array comes from retworkx `distance_matrix()` function) because it's just counting edges between nodes in the graph. So when this is called with `cdist` we don't really have to worry about it, but for the times this is called with `scale` that is actually a float.

Reply:
To be honest, it probably needs a bunch of considerations for the fastest speed here, because there's a couple of levels of memory indirection before we get to the register variables we want.

Yeah, the threading inside process parallelisation may well not be worth it in the end.  Maybe something to look at in the future, but probably not super high priority.

Reply:
To be honest, it probably needs a bunch of considerations for the fastest speed here, because there's a couple of levels of memory indirection before we get to the register variables we want.

Yeah, the threading inside process parallelisation may well not be worth it in the end.  Maybe something to look at in the future, but probably not super high priority.

Reply:
Well we're already multithreaded here. That's where the speed advantage comes in this PR, the outer trials loop is executed in a rayon parallel iterator. This does raise a good point that we might still be in parallel processes too (although given the state of our multiprocessing support on windows, macOS>=py3.8, or linux==py3.9) and this multithreaded single pass might be more harmful for performance then. What I meant here though was nested multithreaded worker pools. I did try this locally just a little while ago to see what would happen and it basically has pegged all 64 threads on my workstation for about 1.5 hours now. So definitely not the approach we want to take. :) (we might be able to try to balance threads a bit like on my workstation I have 44 free threads, when running with 20 trials, which means we could probably get away with using 2 threads per trial here to do this sum in parallel, but my workstation is a pretty extreme case).

I was also looking at [`simdeez`](https://docs.rs/simdeez/latest/simdeez/) as a potential way to use simd here, but yeah I agree it will require more thought (I'll also have to brush up on my vectorized intrinsics to figure out how to use that lib) and is a future thing. We can leave optimizing this function as a future todo